arch/x86/pagetables.c

```diff
@@ -272,6 +273,80 @@ static void map_tmp_mapping_entry(void) {
     _tmp_mapping_entry = paddr_to_virt_kern(_paddr(entry));
 }
 
+static int get_va_mfn_order(const cr3_t *cr3_ptr, const void *va, mfn_t *mfn,
+                            unsigned int *order) {
+    unsigned int _order;
+    mfn_t _mfn;
+    pgentry_t *tab;
+
+    if (mfn_invalid(cr3_ptr->mfn))
+        return -EINVAL;
+
+    dprintk("%s: va: 0x%p (cr3: 0x%p)\n", __func__, va, cr3_ptr);
+
+    tab = tmp_map_mfn(cr3_ptr->mfn);
+#if defined(__x86_64__)
+    pml4_t *l4e = l4_table_entry((pml4_t *) tab, va);
+    if (mfn_invalid(l4e->mfn) || !l4e->P)
+        return -ENOENT;
+
+    tab = tmp_map_mfn(l4e->mfn);
+#endif
+    pdpe_t *l3e = l3_table_entry((pdpe_t *) tab, va);
+    if (l3e->PS) {
+        _mfn = l3e->mfn;
+        _order = PAGE_ORDER_1G;
+        goto done;
+    }
+
+    if (mfn_invalid(l3e->mfn) || !l3e->P)
+        return -ENOENT;
+
+    tab = tmp_map_mfn(l3e->mfn);
+    pde_t *l2e = l2_table_entry((pde_t *) tab, va);
+    if (l2e->PS) {
+        _mfn = l2e->mfn;
+        _order = PAGE_ORDER_2M;
+        goto done;
+    }
+
+    if (mfn_invalid(l2e->mfn) || !l2e->P)
+        return -ENOENT;
+
+    tab = tmp_map_mfn(l2e->mfn);
+    pte_t *l1e = l1_table_entry((pte_t *) tab, va);
+    _mfn = l1e->mfn;
+    _order = PAGE_ORDER_4K;
+
+done:
+    if (mfn)
+        *mfn = _mfn;
+    if (order)
+        *order = _order;
```

Comment:
is it by design to allow null ptrs for mfn and order? Otherwise maybe possible to crash or warn as early as possible if it happens?

Reply:
Main point was to allow either mfn or order pointers or both. We can add an ASSERT when there is none.